### PR TITLE
Correct source code locations of link diagnostics for Objective-C doc comments

### PR DESCRIPTION
**Pull request by patshaughnessy**
**Explanation**: Fix the source location for some Obj-C diagnostics.
**Scope**: Fixes the line and column numbers that appear in the diagnostics.json file, used by Xcode to determine where to show them in the editor.
**GitHub Issue**: none.
**Risk**: Low. Minor code changes impacting diagnostic source locations only..
**Testing**: Ran the ./bin/test script and it succeeded
**Reviewer**: @d-ronnqvist

Bug/issue #, if applicable: 

rdar://102855081

## Summary

Correct the line and character locations of invalid link diagnostics in Objective-C source code doc comments.

Previously, `DocumentationContext#resolveLinks` would use the default symbol to calculate the offset of each doc comment in a source file. However, the default symbol is always set to a Swift symbol even for Objective-C source code. This lead to a bug when Objective-C diagnostics would always use a zero offset.

Example:

Example of an Objective-C link diagnostic at the wrong location

Instead, save the offset of each doc comment in the `DocumentationChunk` structure, which is initialized using the documented symbol, i.e. the Swift or Objective-C symbol which actually contains the doc comment.

Also decrement the line and character indexes by 1 for both Objective-C doc comments so Xcode displays them in the proper location.

Same example with the fix in place:

Example of an Objective-C link diagnostic at the wrong location

## Dependencies

None.

## Testing

Introduced invalid links into doc comments inside of various Objective-C header files, and built the documentation. Tested that Xcode displayed the diagnostics in the proper location in the source code editor.

Steps:
1. Add ``invalid-link`` or some other invalid DocC link in a doc comment in an Objective-C header file.
2. Build the documentation using Xcode.
3. Check that the yellow warning diagnostic appears in the proper location. See the screen shots above.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ x ] Added tests
- [ x ] Ran the `./bin/test` script and it succeeded
- ~~[ ] Updated documentation if necessary~~

### CODE REVIEW

**Comment by patshaughnessy** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/Test Bundles/ObjCFrameworkWithInvalidLink.docc/Info.plist [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Test Bundles/ObjCFrameworkWithInvalidLink.docc/Info.plist:1]: FIY: You don't need an Info.plist unless you're specifying default availability or default syntax highlighting. You can remove this file and the test will work the same.

File: Tests/SwiftDocCTests/Utility/SymbolExtensionsTests.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            )
+            let line = SymbolGraph.LineList.Line(text: str, range: range)
+            return SymbolGraph.LineList([line])
+        }
+
+        func createSymbol(interfaceLanguage: String, docs: String? = nil) -> SymbolGraph.Symbol {
+            return SymbolGraph.Symbol(
+                identifier: .init(precise: "abcd", interfaceLanguage: interfaceLanguage),
+                names: .init(title: "abcd-in-\(interfaceLanguage)", navigator: nil, subHeading: nil, prose: nil),
+                pathComponents: ["documentation", "mykit", "abcd"],
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Utility/SymbolExtensionsTests.swift:35]: minor: Neither "documentation" not the module name are included in the path components that are emitted by Clang or the Swift compiler.

File: Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift [hunk 3]
```diff
@@ -3908,6 +3908,25 @@ let expected = """
...
+        // One doc comment in the Obj-C header file contains an invalid doc
+        // link on line 24, columns 56-63:
+        // "Log a hello world message. This line contains an ``invalid`` link."
+        let (_, _, context) = try testBundleAndContext(copying: "ObjCFrameworkWithInvalidLink")
+        let problems = context.problems
+        XCTAssertEqual(1, problems.count)
+        let problem = try XCTUnwrap(problems.first)
+        let basename = try XCTUnwrap(problem.diagnostic.source?.lastPathComponent)
+        XCTAssertEqual("HelloWorldFramework.h", basename)
+        let start = Markdown.SourceLocation(line: 24, column: 56, source: nil)
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift:3925]: Question: should these source locations have a `source` value? I genuinely don't remember when that's set and when it's not.

File: Sources/SwiftDocC/Utility/SymbolExtensions.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    /// - Objective-C: subtract one from the start/end lines, and the start/end characters
+    /// - Swift and other languages: return the original offset
+    ///
+    /// - Returns: A ``SourceRange`` or nil if this symbol has no doc comment.
+    func offsetAdjustedForInterfaceLanguage() -> SymbolGraph.LineList.SourceRange? {
+        guard let range = docComment?.lines.first?.range else {
+            return nil
+        }
+        // Return the original range for Swift and other languages
+        guard identifier.interfaceLanguage == "objective-c" else {
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Utility/SymbolExtensions.swift:27]: Should we also support source language aliases here?
```suggestion
        guard SourceLanguage(knownLanguageIdentifier: identifier.interfaceLanguage) == .objectiveC else {
```

Comment (d-ronnqvist) [reply Sources/SwiftDocC/Utility/SymbolExtensions.swift:27]: Actually, is there something general we could do here that's not conditional on the source language?

File: Tests/SwiftDocCTests/Test Bundles/ObjCFrameworkWithInvalidLink.docc/Info.plist (hunk 1 above)
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Test Bundles/ObjCFrameworkWithInvalidLink.docc/Info.plist:1]: removed

File: Tests/SwiftDocCTests/Utility/SymbolExtensionsTests.swift (hunk 2 above)
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Utility/SymbolExtensionsTests.swift:35]: makes sense; thx - removed those

File: Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift (hunk 3 above)
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift:3925]: It looks like source is the path/URL of the markdown file, which are parsed by the swift-markdown package. However, in this test and code path, the markdown comes from a doc comment from a Swift code file... in this test one of the symbol graph files under `Test Bundles/ObjCFrameworkWithInvalidLink.docc`. When DocC parses markdown from doc comments, it looks like it leaves the source URL as the default value nil, here in the call to the `Document` initializer:

```
let documentOptions: ParseOptions = [.parseBlockDirectives, .parseSymbolLinks, .parseMinimalDoxygen]
let docCommentMarkup = Document(parsing: docCommentString, options: documentOptions)
```
(from: https://github.com/apple/swift-docc/blob/main/Sources/SwiftDocC/Model/DocumentationNode.swift#L442)

Anyway, this test is about checking the line and column values returned by the call to:

```
let offset = symbol.offsetAdjustedForInterfaceLanguage()
```

...which I added in this PR and immediately follows the first snippet above on my branch. So the value of source isn't relevant to this test anyway.

File: Sources/SwiftDocC/Utility/SymbolExtensions.swift (hunk 4 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Utility/SymbolExtensions.swift:27]: I'll take the suggested change - thanks!

**Comment by patshaughnessy** [comment]
@swift-ci please test

**Review by d-ronnqvist — APPROVED** [review]

**Review by franklinsch — APPROVED** [review]

**Comment by patshaughnessy** [comment]
@swift-ci please test

**Pull request merged**